File: src/lidar/pillarize.py
# Source: Lang et al., "PointPillars: Fast Encoders for Object Detection from Point Clouds," CVPR 2019.
# Converts a raw LiDAR point cloud into a dense pillar tensor ready for PointNet encoding.

import torch


def discretize_point_cloud(points: torch.Tensor, voxel_size: tuple, point_cloud_range: tuple):
    """
    Assign each point to a pillar grid cell.

    points: (N, 4) — x, y, z, intensity
    voxel_size: (x_size, y_size, z_size)
    point_cloud_range: ((x_min, x_max), (y_min, y_max), (z_min, z_max))

    returns: (N, 2) pillar grid indices (ix, iy) per point
    """
    (x_min, x_max), (y_min, y_max), _ = point_cloud_range
    x_size, y_size, _ = voxel_size
    grid_w = round((x_max - x_min) / x_size)
    grid_h = round((y_max - y_min) / y_size)

    # A point can pass the caller's `< x_max` range filter (in float32) yet
    # still floor to exactly grid_w here, since the filter's comparison and
    # this division/floor round differently right at the boundary. Rare, but
    # with hundreds of thousands of points per sample (multi-sweep LiDAR) it
    # does happen — clamp rather than let a downstream scatter go out of bounds.
    ix = ((points[:, 0] - x_min) / x_size).floor().long().clamp(0, grid_w - 1)
    iy = ((points[:, 1] - y_min) / y_size).floor().long().clamp(0, grid_h - 1)

    return torch.stack((ix, iy), dim=1)


def get_pillar_centers(pillar_indices: torch.Tensor, voxel_size: tuple, point_cloud_range: tuple):
    """
    Compute the geometric center (x, y) of each pillar from its grid index.

    pillar_indices: (N, 2) — (ix, iy) per point
    returns: (N, 2) — (x_center, y_center) in ego-frame meters per point
    """
    (x_min, _), (y_min, _), _ = point_cloud_range
    x_size, y_size, _ = voxel_size

    cx = x_min + (pillar_indices[:, 0] + 0.5) * x_size
    cy = y_min + (pillar_indices[:, 1] + 0.5) * y_size

    return torch.stack((cx, cy), dim=1)


def augment_pillars(points: torch.Tensor, pillar_centers: torch.Tensor, cluster_centers: torch.Tensor):
    """
    Augment each point with offset features as described in the PointPillars paper.

    x_c, y_c, z_c — offset from the arithmetic mean of all points in the pillar (cluster centroid)
    x_p, y_p      — offset from the geometric pillar center

    points:          (N, 4) — x, y, z, intensity
    pillar_centers:  (N, 2) — geometric pillar center per point
    cluster_centers: (N, 3) — mean x, y, z of all points in the pillar per point

    returns: (N, 9)
    """
    x_c = points[:, 0] - cluster_centers[:, 0]
    y_c = points[:, 1] - cluster_centers[:, 1]
    z_c = points[:, 2] - cluster_centers[:, 2]

    x_p = points[:, 0] - pillar_centers[:, 0]
    y_p = points[:, 1] - pillar_centers[:, 1]

    return torch.cat([
        points,
        x_c.unsqueeze(1), y_c.unsqueeze(1), z_c.unsqueeze(1),
        x_p.unsqueeze(1), y_p.unsqueeze(1),
    ], dim=1)


def optimize_pillars(points: torch.Tensor, pillar_indices: torch.Tensor, max_points_per_pillar: int):
    """
    Group points into pillars and pad/truncate to a fixed size.

    points:         (N, 9) — augmented point features
    pillar_indices: (N, 2) — (ix, iy) grid index per point
    max_points_per_pillar: fixed pillar size N

    returns:
        pillar_features: (P, N, 9)
        unique_indices:  (P, 2) — (ix, iy) per pillar
    """
    unique_indices, inverse = torch.unique(pillar_indices, dim=0, return_inverse=True)
    P = unique_indices.shape[0]

    # Vectorized per-pillar slot assignment instead of a Python loop over pillars
    # (see _cluster_centers in point_pillars.py for why). Rank = each point's
    # position within its own pillar, in original point order — computed via a
    # stable sort by pillar so equal-pillar points keep their relative order,
    # then position-within-run via cumcount-per-group.
    order          = torch.argsort(inverse, stable=True)
    sorted_inverse = inverse[order]
    is_group_start = torch.ones_like(sorted_inverse, dtype=torch.bool)
    is_group_start[1:] = sorted_inverse[1:] != sorted_inverse[:-1]
    group_start_pos = torch.zeros(P, dtype=torch.long, device=points.device)
    group_start_pos[sorted_inverse[is_group_start]] = torch.nonzero(is_group_start, as_tuple=True)[0]
    rank_sorted = torch.arange(len(sorted_inverse), device=points.device) - group_start_pos[sorted_inverse]
    rank = torch.empty_like(rank_sorted)
    rank[order] = rank_sorted

    keep = rank < max_points_per_pillar
    pillar_features = torch.zeros(P, max_points_per_pillar, points.shape[1],
                                  device=points.device, dtype=points.dtype)
    pillar_features[inverse[keep], rank[keep]] = points[keep]

    return pillar_features, unique_indices

File: src/cpp/pillarize.cpp
#include <map>
#include <cmath>
#include <pillarize.h>

grid_params get_grid_params(voxel_size vs, point_cloud_range range) {
    return {range.x_min, range.y_min, vs.x, vs.y};
}

std::vector<pillar_index> discretize_point_clouds(const std::vector<point>& points,
                                                  voxel_size vs,
                                                  point_cloud_range range) {
    const auto gp = get_grid_params(vs, range);

    std::vector<pillar_index> indices;
    indices.reserve(points.size());
    for (const point& p : points) {
        int ix = static_cast<int>(std::floor((p.x - gp.x_min) / gp.x_size));
        int iy = static_cast<int>(std::floor((p.y - gp.y_min) / gp.y_size));
        indices.push_back({ix, iy});
    }
    return indices;
}

std::vector<pillar_center> get_pillar_centers(const std::vector<pillar_index>& pillar_indices,
                                              voxel_size vs,
                                              point_cloud_range range) {
    const auto gp = get_grid_params(vs, range);

    std::vector<pillar_center> centers;
    centers.reserve(pillar_indices.size());
    for (const pillar_index& p : pillar_indices) {
        float cx = gp.x_min + (p.ix + 0.5f) * gp.x_size;
        float cy = gp.y_min + (p.iy + 0.5f) * gp.y_size;
        centers.push_back({cx, cy});
    }
    return centers;
}

std::vector<augmented_point> augment_pillars(const std::vector<point>& points,
                                             const std::vector<pillar_center>& pillar_centers,
                                             const std::vector<cluster_center>& cluster_centers) {
    std::vector<augmented_point> augmented_points;
    augmented_points.reserve(points.size());
    for (size_t i = 0; i < points.size(); i++) {
        augmented_point ap;
        ap.x         = points[i].x;
        ap.y         = points[i].y;
        ap.z         = points[i].z;
        ap.intensity = points[i].intensity;

        ap.x_cluster_offset = points[i].x - cluster_centers[i].x;
        ap.y_cluster_offset = points[i].y - cluster_centers[i].y;
        ap.z_cluster_offset = points[i].z - cluster_centers[i].z;

        ap.x_pillar_offset = points[i].x - pillar_centers[i].x;
        ap.y_pillar_offset = points[i].y - pillar_centers[i].y;
        augmented_points.push_back(ap);
    }
    return augmented_points;
}

pillar_batch optimize_pillars(const std::vector<augmented_point>& points,
                              const std::vector<pillar_index>& pillar_indices,
                              int max_points_per_pillar) {
    std::map<std::pair<int, int>, int> pillar_map;
    std::vector<int> inverse(pillar_indices.size());

    for (size_t i = 0; i < pillar_indices.size(); i++) {
        auto key = std::make_pair(pillar_indices[i].ix, pillar_indices[i].iy);
        if (pillar_map.find(key) == pillar_map.end()) {
            int slot = static_cast<int>(pillar_map.size());
            pillar_map[key] = slot;
        }
        inverse[i] = pillar_map[key];
    }

    int P = static_cast<int>(pillar_map.size());
    std::vector<std::vector<augmented_point>> pillar_features(P);

    for (size_t i = 0; i < points.size(); i++) {
        int slot = inverse[i];
        if (pillar_features[slot].size() < static_cast<size_t>(max_points_per_pillar)) {
            pillar_features[slot].push_back(points[i]);
        }
    }

    std::vector<pillar_index> unique_indices(P);
    for (const auto& [key, slot] : pillar_map) {
        unique_indices[slot] = {key.first, key.second};
    }
    return {pillar_features, unique_indices};
}